bundler/spec/bundler/definition_spec.rb

```diff
@@ -277,6 +277,59 @@
     end
   end
 
+  describe "#precompute_source_requirements_for_indirect_dependencies?" do
+    before do
+      allow(Bundler::SharedHelpers).to receive(:find_gemfile) { Pathname.new("Gemfile") }
+    end
+
+    let(:sources) { Bundler::SourceList.new }
+    subject { Bundler::Definition.new(nil, [], sources, []) }
+
+    before do
+      subject.instance_variable_set(:@remote, true)
+      allow(sources).to receive(:aggregate_global_source?).and_return(false)
+      allow(sources).to receive(:non_global_rubygems_sources).and_return(non_global_rubygems_sources)
+    end
+    context "when remote and does not have multiple global_source" do
+      context "when all the scoped sources contain a dependency API" do
+        let(:non_global_rubygems_sources) do
+          [
+            double("non-global-source-0", :dependency_api_available? => true, :to_s => "a"),
+            double("non-global-source-1", :dependency_api_available? => true, :to_s => "b"),
+          ]
+        end
+
+        it "will not raise a warning" do
+          expect(subject.send(:precompute_source_requirements_for_indirect_dependencies?)).to be_truthy
+          expect(subject).not_to receive(:non_dependency_api_warning)
+        end
+      end
+
+      context "when scoped sources do not contain a dependency API" do
+        let(:non_global_rubygems_sources) do
+          [
+            double("non-global-source-0", :dependency_api_available? => true, :to_s => "a"),
+            double("non-global-source-1", :dependency_api_available? => false, :to_s => "b"),
+            double("non-global-source-2", :dependency_api_available? => false, :to_s => "c"),
+          ]
+        end
+
+        it "will raise a warning" do
+          expect(subject.send(:precompute_source_requirements_for_indirect_dependencies?)).to be_falsy
+          expect(subject).to receive(:non_dependency_api_warning)
```

Comment:
@deivid-rodriguez I added the tests. But the test fail in this line. RSpec 3's `receive` method doesn't count as expected. Do you know what's wrong in my code?

```
$ bin/rspec spec/bundler/definition_spec.rb:322

  1) Bundler::Definition#precompute_source_requirements_for_indirect_dependencies? when remote and does not have multiple global_source when scoped sources do not contain a dependency API will raise a warning
     Failure/Error: expect(subject).to receive(:non_dependency_api_warning)
...
           expected: 1 time with any arguments
           received: 0 times with any arguments
```

Reply:
Not sure, I think you need to setup the expectation before actually calling the method, or use `have_received` if you do it after?

Reply:
@deivid-rodriguez I found the reason. As you said, the expectation logic should be before the actual calling. and not only that, for example the following code don't work at the line `expect(Bundler.ui).to receive(:warn)`. Because the first line `expect(subject).to receive(:non_dependency_api_warning)` hides the received count of nested method `Bundler.ui.warn` actually called in the `non_dependency_api_warning`. So, we cannot use the `expect`, `receive` for the nested 2 methods in the one `it` block.

```
expect(subject).to receive(:non_dependency_api_warning)
expect(Bundler.ui).to receive(:warn) # <= can't count the received method in the `non_dependency_api_warning`.
expect(subject.send(:precompute_source_requirements_for_indirect_dependencies?)).to be_falsy
```